### ISSUE TITLE
Replacing a related child entity while also dropping the old child entity

### ISSUE THREAD

**Issue by poke** [aspnet/EntityFrameworkCore#13219]
I am having a problem that I want to replace a child entity on an object with a new entity. While doing that, I want to replace the reference on the parent entity, create a new child entity, and *also* drop the original/old child entity.

When attempting to do that, even if explicitly dropping the old entity, I am unable to set the *new* entity within the same transaction. I am always left with either a dangling old or a dangling new child entity.

This is very likely related to #13151, or possibly even the same, but I do have a simple and straight-forward way to reproduce this so you can try it yourself:

```c#
public class Parent
{
    public int Id { get; set; }
    public int? ChildId { get; set; }
    public Child Child { get; set; }
}

public class Child
{
    public int Id { get; set; }
    public Parent Parent { get; set; }
}

public class TestContext : DbContext
{
    public DbSet<Parent> Parents { get; set; }
    public DbSet<Child> Children { get; set; }
    public TestContext(DbContextOptions options) : base(options) { }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Parent>(b =>
        {
            b.HasOne(p => p.Child).WithOne(c => c.Parent)
                .HasForeignKey<Parent>(p => p.ChildId);
        });
    }
}
```

This is the test code (using xunit asserts):

```c#
var dbOptions = new DbContextOptionsBuilder().UseInMemoryDatabase("Test").Options;

using (var db = new TestContext(dbOptions))
{
    var parent = new Parent { Id = 1, Child = new Child { Id = 2 } };
    db.Add(parent);
    db.SaveChanges();
}

using (var db = new TestContext(dbOptions))
{
    var parent = db.Parents.Include(p => p.Child).First(p => p.Id == 1);

    // Case 1: Ideally, just replace the child
    parent.Child = new Child { Id = 3 };

    // Case 2: Remove old child first
    db.Remove(parent.Child);
    parent.Child = new Child { Id = 3 };

    // Case 3: Save after removal
    db.Remove(parent.Child);
    db.SaveChanges();
    parent.Child = new Child { Id = 3 };

    db.SaveChanges();
    Assert.Equal(3, parent.ChildId);
}

using (var db = new TestContext(dbOptions))
{
    var parent = db.Parents.Find(1);
    Assert.Equal(3, parent.ChildId);

    var oldChild = db.Children.Find(2);
    Assert.Null(oldChild);
    var newChild = db.Children.Find(3);
    Assert.NotNull(newChild);
}
```

As you can see, I have three different test cases in there. You can comment them in or out to test the different ideas:

1. Ideally, I would just want to replace the child entity on the parent object with a new one and have the old one automatically dropped. But I wouldn’t actually expect this to work.
2. Case 2 is what I would expect and want to do. Remove the old child and then just set a new one. However, this does not work because apparently, EF still remembers that I am dropping the old child entity and as such executes the `OnDelete` action to set the `parent.ChildId` to `null`. This even happens when I set it explicitly to `3`.

   So basically, I am completely unable to set a new child value other than `null` if I remove the old child entity.
3. Case 3 is what obviously works. Using separate transactions, I can first remove the original child, and then add a new one in another transaction. But that really isn’t what I want to do since I want to make this transactional safe.

**Comment by poke** [comment]
Case 4:

```c#
parent.Child.Parent = null;
db.Remove(parent.Child);
parent.Child = new Child { Id = 3 };
```

This also works, since when deleting the child, EF is now unable to navigate back to the parent to set the `ChildId` to `null`.

**Comment by ajcvickers** [comment]
@poke What version of EF Core are you using?

**Comment by poke** [comment]
@ajcvickers I’m running above test code inside LINQPad with EF Core 2.1.2. I can provide a clonable repo though if needed.

**Comment by ajcvickers** [comment]
@poke It's quite unusual to have the FK property defined on the "Parent" side of the relationship. Usually when people talk about Parent/Child relationships the Parent acts as the principal side of the relationship and the Child acts as the dependent side. So, for example, cascade deletes always work from the principal to the dependent, not the other way round.

That being said, when I run your code I see the following:
- Case 1: This case creates a new principal, then assigns the existing dependent to the new principal--this is typically known as re-parenting. (Although with child and parent switched in your model maybe it should be called re-childing in this case? ;-) ) This is done at the FK level by changing the FK of the child to point to the new principal. The old principal is not changed and remains in the database.
- Case 2: This works for me--the existing principal is first deleted, and the child has it's FK set to point to the new principal.
- Case 3: This works for me too.

Everything here is as expected. For case 2, can you post a runnable project/solution  or complete code listing that demonstrates that the existing principal is not deleted?

**Comment by ajcvickers** [comment]
@poke I take that back. Case 2 doesn't work--the existing principal is deleted, but the child FK is not set to the new principal. I'll look into this further.

**Comment by poke** [comment]
Yeah, it could be true that this is a bit unusual. I’m originally encountering this issue in a situation where “child” is a file item that can be attached to multiple different kind of objects. Having the FK on the “parent” allows me to reuse this for different types of parents. And it also gives me a nice shortcut to figure out whether a child/file exists without having to query for it.

But if it makes more sense to you if you relabel those two classes (or call them `A` and `B`) – sure! :smile:

But yeah, the issue with case 2 is that the foreign key is not set, or probably reset again by applying the `OnDelete` trigger.
